src/common/rabbitmq/rabbitmq.service.ts

```diff
@@ -0,0 +1,116 @@
+import { Injectable, Logger, OnModuleDestroy } from "@nestjs/common";
+import { ConfigService } from "@nestjs/config";
+
+/**
+ * Service for publishing messages to a RabbitMQ queue.
+ */
+@Injectable()
+export class RabbitMQService implements OnModuleDestroy {
+  private connectionOptions: amqp.Options.Connect;
+  private connection: Connection;
+  private channel: Channel;
+
+  constructor(private configService: ConfigService) {
+    Logger.log("Initializing RabbitMQService.", "RabbitMQService");
+
+    const rabbitMqEnabled =
+      this.configService.get<string>("rabbitMq.enabled") === "yes";
+
+    if (rabbitMqEnabled) {
+      const hostname = this.configService.get<string>("rabbitMq.hostname");
+      const port = this.configService.get<number>("rabbitMq.port");
+      const username = this.configService.get<string>("rabbitMq.username");
+      const password = this.configService.get<string>("rabbitMq.password");
+
+      if (!hostname || !port || !username || !password) {
+        Logger.error(
+          "RabbitMQ is enabled but missing one or more config variables.",
+          "RabbitMQService",
+        );
+      } else {
+        this.connectionOptions = {
+          protocol: "amqp",
+          hostname: hostname,
+          port: port,
+          username: username,
+          password: password,
+        };
+
+        amqp.connect(
+          this.connectionOptions,
+          (connectionError: Error, connection: Connection) => {
+            if (connectionError) {
+              Logger.error(
+                "Connection error in RabbitMQService: " +
+                  JSON.stringify(connectionError.message),
+                "RabbitMQService",
+              );
+              return;
+            }
+            this.connection = connection;
+
+            this.connection.createChannel(
+              (channelError: Error, channel: Channel) => {
+                if (channelError) {
+                  Logger.error(
+                    "Channel error in RabbitMQService: " +
+                      JSON.stringify(channelError.message),
+                    "RabbitMQService",
+                  );
+                  return;
+                }
+                this.channel = channel;
+                Logger.log(this.channel);
+              },
+            );
+          },
+        );
+      }
+    } else {
+      Logger.error("RabbitMQ is not enabled.", "RabbitMQService");
+    }
+  }
+
+  connect(queue: string, exchange: string, key: string) {
+    try {
+      this.channel.assertQueue(queue, { durable: true });
+      this.channel.assertExchange(exchange, "topic", {
+        durable: true,
+      });
+      this.channel.bindQueue(queue, exchange, key);
+    } catch (error) {
+      Logger.error(
+        `Could not connect to RabbitMQ queue ${queue} with exchange ${exchange} and key ${key}.`,
+        "RabbitMQService",
+      );
+      Logger.error(error);
+    }
+  }
+
+  sendMessage(queue: string, message: string) {
```

Comment:
What happens if a message is sent without first calling connect? If they are always called in pairs would it make more sense to merge them into one method?